cpuid-bool/src/lib.rs

```diff
@@ -1,51 +1,118 @@
 //! Macro for checking CPU capabilities at runtime.
 //!
-//! # Usage example
+//! # Example
 //! ```
-//! if cpuid_bool::cpuid_bool!("sha", "aes") {
+//! // This macro creates `cpuid_aes_sha` module
+//! cpuid_bool::new!(cpuid_aes_sha, "aes", "sha");
+//!
+//! // `token` is a Zero Sized Type value, which guarantees
+//! // that underlying static storage got properly initialized,
+//! // which allows to omit initialization branch
+//! let token: cpuid_aes_sha::InitToken = cpuid_aes_sha::init();
+//! if token.get() {
 //!     println!("CPU supports both SHA and AES extensions");
 //! } else {
 //!     println!("SHA and AES extensions are not supported");
 //! }
+//!
+//! // If stored value needed only once you can get stored value
+//! // omitting the token
+//! let val = cpuid_aes_sha::get();
+//! assert_eq!(val, token.get());
+//!
+//! // Additionally you can get both token and value
+//! let (token, val) = cpuid_aes_sha::init_get();
+//! assert_eq!(val, token.get());
 //! ```
 //! Note that if all tested target features are enabled via compiler options
-//! (e.g. by using `RUSTFLAGS`), `cpuid_bool!` macro immideatly will expand
-//! to `true` and will not use CPUID instruction. Such behavior allows
-//! compiler to eliminate fallback code.
+//! (e.g. by using `RUSTFLAGS`), the `get` method will always return `true`
+//! and `init` will not use CPUID instruction. Such behavior allows
+//! compiler to completely eliminate fallback code.
```

Comment:
I was a bit confused by this section (which I guess was there to begin with). It would seem to defeat the point of having runtime detection at all?

One of the main use cases I have in mind is when compiled binaries are relocated to environments where one or more of the features are unavailable, e.g. a VM which does not support AES-NI.

Reply:
Programs compiled with enabled target features are not fully portable in principle. If you enable non-default target features (e.g. AVX2 or AVX512), compiler in its full right to use those instructions to optimize code. Thus the resulting binary would not work in environments which do not support the enabled extension. If you can not guarantee that program will not be executed only on CPUs with AES-NI support you should not enable the `aes` target feature via compiler flags but if you can not guarantee it, then by enabling it you will get a bit more efficient binary (smaller size of the AES types and lack of fallback branches).

I am not happy with the current description, so I will gladly replace it with a better explanation if you have one.

Reply:
To break the problem down as I understand it:

- In order for Rust to emit AES-NI instructions, the `target_feature` for `aes` *MUST* be enabled. If the target feature is not enabled, LLVM will use a software implementation of the AES-NI instructions (see [discussion here](https://github.com/RustCrypto/block-ciphers/pull/208#issuecomment-735400697))
- Even on the same *physical* CPU, it's possible for AES-NI to be enabled or disabled, depending on whether AES-NI is enabled or disabled in the BIOS, or whether the workload is virtualized and virtualization supports or does not support AES-NI. This is a bit different from other CPU instructions which we could otherwise assume are always available given some CPU microarchitecture as a baseline.

So I think the approach in this PR can work, but we'll need e.g. the `aes` target feature enabled at the time the program is compiled, and for `cpuid_bool` to always perform the check regardless of enabled target features.

Reply:
Never mind, I was wrong.